text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+
+As well as metadata based on types and names, the oracle should keep track of
+warnings. Since code with warnings but no errors is not re-compiled, a tool
+outside the compiler must track them for display in the IDE. This will be done
+by the oracle.
+
+
+### Details
+
+#### API
+
+The oracle's API is a set of IPC calls. How exactly these should be implemented
+is not clear. The most promising options are sending JSON over TCP, using
+[thrift](https://thrift.apache.org/), or using Cap'n Proto (I'm unclear about
+exactly what the transport layer looks like using Cap'n Proto, there is no Cap'n
+Proto RCP implementation for Java, but I believe there is an alternative using
+shared, memory mapped files as a buffer; I'm not familiar enough with the
+library to work out what is needed).
+
+I've detailed the API I believe we'll need to start with. This is slightly more
+than a minimal set. I expect it will expand as time goes by. At some point we
+will want to stabilise parts of the API to allow for third party implementations
+of the oracle and compiler.
+
+All API calls can return success or error results. Many calls involve a *span*;
+for the oracle's API, this is defined as two byte offsets from the start of the
+file (oracle spans must always be contained in a single file).
+
+There are some alternative span definitions: we could use file and column indices
+rather than byte offsets (this has some edge case difficulties with the
+definition of a newline - do unicode newlines count? It also requires some extra
+computation), we could use character offsets (again involves some more
+computation, but might be more robust).
+
+A problem is that Visual Studio uses UTF16 while Rust uses UTF8, there is (I
+understand) no efficient way to convert between byte counts in these systems.
+I'm not sure how to address this. It might require the oracle to be able to
+operate in UTF16 mode.
+
+Where no return value is specified, the call returns success or failure (with a
+reason).
+
+The philosophy of the API is that most functions should only take a single call,
+as opposed to making each function as minimal and orthogonal as possible. This
+is because IPC can be slow and response time is important for IDEs.
+
+
+**Projects**
+
+Note that the oracle stores no metadata about a project.
+
+*init project*
+
+Takes a project name, returns an id string (something close to the project's name).
+
+*delete project*
+
+Takes a project id.
+
+*list projects*
+
+Takes nothing, returns a list of project ids.
+
+Each of the remaining calls takes a project identifier.
+
+
+**Update**
+
+See section on input data format below.
+
+*update*
+
+Takes input data (actual source code rather than spans since we cannot assume
+the user has saved the file) and a list of spans to invalidate.  Where there are
+no invalidated spans, the update call adds data (which will cause an error if
+there are conflicts). Where there is no input data, update just invalidates.
+
```

Comment:
I understand this is not a detailed API, but I dont see how this works when there are multiple spans to invalidate.Do all spans must be computed based on the initial content of the file? It sounds like it could be a bit painful to compute it from the plugin side. Maybe a simpler API could live alongside this one, that takes the full content of the file and that lets the oracle compute the differences based on its current state.